rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp

```diff
@@ -66,18 +67,57 @@ class TestNodeGraph : public ::testing::Test
 
   const rclcpp::node_interfaces::NodeGraph * node_graph() const {return node_graph_;}
 
+  size_t get_num_topics()
+  {
+    constexpr std::chrono::milliseconds timeout(100);
+
+    int tries = 0;
+    size_t num_topics = 0;
+    while (num_topics < 1u && tries++ < 5) {
+      auto event = node()->get_graph_event();
+      EXPECT_NO_THROW(node()->wait_for_graph_change(event, timeout));
+
+      auto topic_names_and_types = node_graph()->get_topic_names_and_types();
+      num_topics = topic_names_and_types.size();
```

Comment:
Doesn't this mean that you will always wait 100ms in the event that all the graph events occurred before you get to this point?

Maybe it should be more of a while true, with a break when the `num_topics` meets the threshold? And then make that check before waiting?

Reply:
> Doesn't this mean that you will always wait 100ms in the event that all the graph events occurred before you get to this point?

Yep, exactly.  In practice I didn't see this happening, and 100ms was short enough that I didn't think it would be too much of an issue if it did (the test would just be a bit slower).  But I can change it to a different loop to avoid the sleep completely.

Reply:
Well, you don't _have_ to change it, I was just wondering if it was intentional. Especially because it must have worked for fastrtps before this, so in that case we've didn't need the sleep or the retries I guess?

Reply:
> Especially because it must have worked for fastrtps before this, so in that case we didn't need the sleep or the retries I guess?

Yeah, I'm currently thinking that there is some detail of CycloneDDS that makes the graph change just a bit slower.  In practice, this was an extreme flake even with CycloneDDS; I couldn't get it happen with 2 hours of testing.  I could only reproduce it with a lot of load on the machine.

I'm going to go ahead and make the change anyway.  It is a little nicer, and while I'm in there I'll also make the change that @hidmic suggested of using a predicate.